dpctl/tensor/_copy_utils.py

```diff
@@ -763,44 +763,67 @@ def _nonzero_impl(ary):
 
 def _take_multi_index(ary, inds, p):
     if not isinstance(ary, dpt.usm_ndarray):
-        raise TypeError
+        raise TypeError(
+            f"Expecting type dpctl.tensor.usm_ndarray, got {type(ary)}"
+        )
+    ary_nd = ary.ndim
+    p = normalize_axis_index(operator.index(p), ary_nd)
     queues_ = [
         ary.sycl_queue,
     ]
     usm_types_ = [
         ary.usm_type,
     ]
-    if not isinstance(inds, list) and not isinstance(inds, tuple):
+    if not isinstance(inds, (list, tuple)):
         inds = (inds,)
-    all_integers = True
     for ind in inds:
+        if not isinstance(ind, dpt.usm_ndarray):
+            raise TypeError("all elements of `ind` expected to be usm_ndarrays")
         queues_.append(ind.sycl_queue)
         usm_types_.append(ind.usm_type)
-        if all_integers:
-            all_integers = ind.dtype.kind in "ui"
+        if ind.dtype.kind not in "ui":
+            raise IndexError(
+                "arrays used as indices must be of integer (or boolean) type"
+            )
+    res_usm_type = dpctl.utils.get_coerced_usm_type(usm_types_)
     exec_q = dpctl.utils.get_execution_queue(queues_)
     if exec_q is None:
-        raise dpctl.utils.ExecutionPlacementError("")
-    if not all_integers:
-        raise IndexError(
-            "arrays used as indices must be of integer (or boolean) type"
+        raise dpctl.utils.ExecutionPlacementError(
+            "Can not automatically determine where to allocate the "
+            "result or performance execution. "
+            "Use `usm_ndarray.to_device` method to migrate data to "
+            "be associated with the same queue."
         )
     if len(inds) > 1:
+        ind_dt = dpt.result_type(*inds)
+        # ind arrays have been checked to be of integer dtype
+        if ind_dt.kind not in "ui":
+            raise ValueError(
+                "cannot safely promote indices to an integer data type"
+            )
+        inds = tuple(
+            map(
+                lambda ind: ind
+                if ind.dtype == ind_dt
+                else dpt.astype(ind, ind_dt),
+                inds,
+            )
+        )
         inds = dpt.broadcast_arrays(*inds)
-    ary_ndim = ary.ndim
-    p = normalize_axis_index(operator.index(p), ary_ndim)
-
-    res_shape = ary.shape[:p] + inds[0].shape + ary.shape[p + len(inds) :]
-    res_usm_type = dpctl.utils.get_coerced_usm_type(usm_types_)
+    ind0 = inds[0]
+    ary_sh = ary.shape
+    p_end = p + len(inds)
+    inds_sz = ind0.size
+    if 0 in ary_sh[p:p_end] and inds_sz != 0:
```

Comment:
```suggestion
    if 0 in ary_sh[p:p_end] and ind0.size != 0:
```

This way computation behind `.size` would only trigger if indexed array is empty.